content/blog/dont-use-preludes-and-globs/index.md

```diff
@@ -0,0 +1,152 @@
+
+[^1]: A glob import (also called a wildcard import) is an import that brings all items of a module into scope. For example, `use std::collections::HashMap;` vs `use std::collections::*;`. 
+
+## I don't like preludes
+
+Preludes help beginners start quickly with zero boilerplate, but can cause namespace pollution and naming conflicts for experienced developers in larger projects.
+
+I have a hard time reviewing your code which uses preludes: looking up a definition might not work in my code review tool, and I'd have to manually track down where a types came from.
+
+Explicit imports bring clarity.
+
+Speaking of which! A good compromise between convenience and clarity is to import whole modules instead of individual items, such as `use std::fmt`. Then you can say `fmt::Debug` instead of `std::fmt::Debug`.
+
+Unless you write a highly critical framework which provides well-established types, which are *absolutely positively always* mandatory in every (or at least most) interactions with your library, don't add a prelude. Even then, make sure that prelude carries its own weight. [Tokio removed their prelude because that wasn't the case](https://github.com/tokio-rs/tokio/issues/3257).
+
+An explicit import might not be that bad. In fact, it might even clear things up for your users, because it's easier to see where imports come from.
+
+If you're worried about ergonomics, why not re-export common types in root? Then they can get imported with `use mycrate::SomeType;` instead of `use mycrate::mymodule::SomeType;`.
+
+```rust
+// Re-export a type in the your lib.rs
+pub use mymodule::SomeType;
+```
+
+On the same note, avoid glob imports (like `use prelude::*`
+or `use mycrate::mymodule::*`).
+The reasons are the same: they hide dependencies and make it harder to grep for usages of a type or function.
+
+## Common Arguments for Preludes And Glob Imports
+
+Let's address some common arguments in favor of preludes and glob imports
+and see if there's a better way to achieve the same goal.
+
+### Preludes in examples
+
+Some people like to use preludes in their documentation examples.
+
+This can be sweet when you quickly want to show how to use your library without too much boilerplate. But at what cost? At what terrible cost?[^2]
+
+[^2]: I'm exaggerating. Bear with me.
+
+Users *can and will* start to copy-paste that prelude into their code, at which point you'll have to support it. It's convenient in the beginning, but can result in a maintenance burden in the long run.
+
+Instead, why not use hidden code lines for imports in your documentation examples? This way, the examples are still easy to read, but users can't copy-paste the imports:
+
+```rust
+//! # Examples
+//!
+//! ```
+//! # use mycrate::SomeType;
+//! # fn main() {
+//! let t = SomeType::new(); 
+//! # }
+//! ```
+```
+
+This will render as:
+
+```rust
+let t = SomeType::new();
+```
```

Comment:
That's a nice trick, but now what? Now you have bad docs! I as a user can't use your example code. When I copy-paste it, it doesn't compile. Two better ways to avoid this problem:

1. Better imports — if you have to hide and ugly mess of `use` statements, you have a bad designed crate. Refactor!
2. More focused examples — show-case one thing and one thing only. The interaction between two or three things should not necessitate more than two or three imports from the same module. Only caveat is a "full" example in the crate docstring. And there we _really_ need those to be copy-pastable for new users.

The only place where you can hide a `use mylib::MyType` without detriment is in the documentation of `MyType`, and afaik you don't need to do that because for the docs generator that's already in-scope.

Reply:
That's right. Changed the wording to highlight these points. Thanks!